src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -0,0 +1,400 @@
+    segment_off_t length,
+    ool_placement_hint_t hint = ool_placement_hint_t::NONE) {
+    // only logical extents should fall in this path
+    assert(is_logical_type(type));
+    auto dtype = get_allocator_type(hint);
+    CachedExtentRef extent;
+    // for extents that would be stored in NVDIMM/PMEM, no delayed
+    // allocation is needed
+    if (need_delayed_allocation(dtype)) {
+      // set a unique temperary paddr, this is necessary because
+      // transaction's write_set is indexed by paddr
+      extent = cache.alloc_new_extent_by_type(t, type, length, true);
+    } else {
+      extent = cache.alloc_new_extent_by_type(t, type, length);
+    }
+    extent->backend_type = dtype;
+    extent->hint = hint;
+    return extent;
+  }
+
+  /**
+   * delayed_alloc_or_ool_write
+   *
+   * Performs any outstanding ool writes and updates pending lba updates
+   * accordingly
+   */
+  using alloc_paddr_iertr = ExtentOolWriter::write_iertr;
+  alloc_paddr_iertr::future<> delayed_alloc_or_ool_write(
+    Transaction& t) {
+    return seastar::do_with(
+      std::map<ExtentAllocator*, std::list<LogicalCachedExtentRef>>(),
+      std::list<std::pair<paddr_t, LogicalCachedExtentRef>>(),
+      [this, &t](auto& alloc_map, auto& inline_list) mutable {
+      auto& alloc_list = t.get_delayed_alloc_list();
+      for (auto& extent : alloc_list) {
+        // extents may be invalidated
+        if (!extent->is_valid()) {
+          continue;
+        }
+        if (should_be_inline(extent)) {
+          auto old_addr = extent->get_paddr();
+          cache.mark_delayed_extent_inline(t, extent);
+          inline_list.emplace_back(old_addr, extent);
+          continue;
+        }
+        auto& allocator_ptr = get_allocator(extent->backend_type, extent->hint);
+        alloc_map[allocator_ptr.get()].emplace_back(extent);
+      }
+      return trans_intr::do_for_each(alloc_map, [&t](auto& p) {
+        auto allocator = p.first;
+        auto& extents = p.second;
+        return allocator->alloc_ool_extents_paddr(t, extents);
+      }).si_then([&inline_list, this, &t] {
+        return trans_intr::do_for_each(inline_list, [this, &t](auto& p) {
+          auto old_addr = p.first;
+          auto& extent = p.second;
+          return lba_manager.update_mapping(
+            t,
+            extent->get_laddr(),
+            old_addr,
+            extent->get_paddr());
+        });
+      });
+    });
+  }
+
+  void add_allocator(device_type_t type, ExtentAllocatorRef&& allocator) {
+    allocators[type].emplace_back(std::move(allocator));
+  }
+
+private:
+  device_type_t get_allocator_type(ool_placement_hint_t hint) {
+    return device_type_t::SEGMENTED;
+  }
+
+  bool should_be_inline(LogicalCachedExtentRef& extent) {
+    return (std::rand() % 2) == 0;
+  }
+
+  ExtentAllocatorRef& get_allocator(
+    device_type_t type,
+    ool_placement_hint_t hint) {
+    auto& devices = allocators[type];
+    return devices[std::rand() % devices.size()];
```

Comment:
same here, if we don't use hint at current stage, add comment //todo to be reminder.